OpenRA.Mods.Common/Traits/Air/Aircraft.cs

```diff
@@ -868,9 +857,51 @@ public Activity MoveTo(CPos cell, Actor ignoreActor, Color? targetLineColor = nu
 				initialTargetPosition, targetLineColor);
 		}
 
-		public Activity MoveIntoWorld(Actor self, CPos cell, SubCell subCell = SubCell.Any)
+		public Activity MoveIntoWorld(Actor self, int delay = 0)
+		{
+			return new MoveIntoWorldActivity(self, delay);
+		}
+
+		class MoveIntoWorldActivity : Activity
 		{
-			return new Fly(self, Target.FromCell(self.World, cell, subCell));
+			readonly Actor self;
+			readonly Aircraft aircraft;
+			readonly int delay;
+
+			public MoveIntoWorldActivity(Actor self, int delay = 0)
+			{
+				this.self = self;
+				aircraft = self.Trait<Aircraft>();
+				IsInterruptible = false;
+				this.delay = delay;
+			}
+
+			protected override void OnFirstRun(Actor self)
+			{
+				if (delay > 0)
+					QueueChild(new Wait(delay));
+
+				var host = aircraft.GetActorBelow();
+				if (host == null)
+					return;
+
+				aircraft.MakeReservation(host);
+
+				if (!aircraft.Info.TakeOffOnCreation)
+					return;
+
+				// Move to the host's rally point if it has one
+				var rp = aircraft.ReservedActor != null ? aircraft.ReservedActor.TraitOrDefault<RallyPoint>() : null;
+				if (self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition).Length <= aircraft.LandAltitude.Length)
+				{
+					if (rp != null)
+						QueueChild(new TakeOff(self, Target.FromCell(self.World, rp.Location)));
```

Comment:
If the `Wait` above is too long `rp.Location` may change between the time this is queued and the activity actually ticks. It would be better to defer this until the first `Tick`, but we'll need to be careful that the host hasn't died in the meantime.